File: gui/src/main/java/nisui/gui/MainFrame.java
package nisui.gui;

import java.io.InputStream;
import java.io.PrintStream;

import javax.swing.JButton;
import javax.swing.JFrame;
import javax.swing.JMenu;
import javax.swing.JMenuItem;
import javax.swing.JMenuBar;
import javax.swing.JTabbedPane;

import nisui.core.NisuiFactory;

public class MainFrame extends JFrame {
	final NisuiFactory nisuiFactory;
	final InputStream in;
	final PrintStream out;

	private JTabbedPane tabbedPane;

	private JMenuBar menuBar;

	public MainFrame(NisuiFactory nisuiFactory, InputStream in, PrintStream out) {
		super("Nisui");
		this.nisuiFactory = nisuiFactory;
		this.in = in;
		this.out = out;
		setDefaultCloseOperation(JFrame.DISPOSE_ON_CLOSE);

		menuBar = new JMenuBar();
		this.setJMenuBar(menuBar);

		addMenuItem("Query", () -> {
			tabbedPane.addTab("Queries", new QueriesPanel(this));
		});

		tabbedPane = new JTabbedPane();
		setContentPane(tabbedPane);
	}

	void addMenuItem(String caption, Runnable action) {
		JMenuItem menuItem = new JMenuItem(caption);
		menuItem.addActionListener(e -> action.run());
		menuBar.add(menuItem);
	}
}

File: gui/src/main/kotlin/nisui/gui/MainFrame.kt
package nisui.gui;

import java.io.InputStream;
import java.io.PrintStream;

import javax.swing.JButton;
import javax.swing.JFrame;
import javax.swing.JMenu;
import javax.swing.JMenuItem;
import javax.swing.JMenuBar;
import javax.swing.JTabbedPane;
import javax.swing.KeyStroke;
import java.awt.event.ActionEvent;
import java.awt.event.KeyEvent;

import nisui.core.NisuiFactory;

import nisui.gui.queries.PlotsPanel;

public class MainFrame(val nisuiFactory: NisuiFactory, val sin: InputStream, val sout: PrintStream): JFrame("Nisui") {
    val menuBar = JMenuBar();
    val tabbedPane = JTabbedPane();

    init {
        nisuiFactory.createResultsStorage().prepareStorage()
        setDefaultCloseOperation(JFrame.DISPOSE_ON_CLOSE);
        setJMenuBar(menuBar);

        addMenuItem("Plot", KeyEvent.VK_P) {
            tabbedPane.addTab("Plot", PlotsPanel(nisuiFactory));
        }

        setContentPane(tabbedPane);
    }

    fun addMenuItem(caption: String, altShortcutKey: Int? = 0, action: () -> Unit): JMenuItem {
        val menuItem = JMenuItem(caption);
        menuItem.addActionListener {action()};
        if (altShortcutKey != null) {
            menuItem.setAccelerator(KeyStroke.getKeyStroke(altShortcutKey, ActionEvent.ALT_MASK));
            menuItem.setMnemonic(altShortcutKey);
        }
        menuBar.add(menuItem);
        return menuItem;
    }
}
